dht_net.go

```diff
@@ -266,49 +288,165 @@ func (ms *messageSender) SendRequest(ctx context.Context, pmes *pb.Message) (*pb
+			}
+		}
+
+		rcount := ms.rcount
+
+		ms.lk.Unlock()
+
+		rctx, cancel := context.WithTimeout(ctx, dhtReadMessageTimeout)
+		defer cancel()
+
+		var res requestResult
+		select {
+		case res = <-resch:
+
+		case <-rctx.Done():
+			return nil, rctx.Err()
+
+		case <-ms.dht.ctx.Done():
+			return nil, ms.dht.ctx.Err()
+		}
+
+		if res.err != nil {
+			ms.lk.Lock()
+			ms.resetSoft(rcount)
+			ms.lk.Unlock()
 
 			if retry {
-				log.Info("error reading message, bailing: ", err)
-				return nil, err
+				log.Info("error reading message, bailing: ", res.err)
+				return nil, res.err
 			} else {
-				log.Info("error reading message, trying again: ", err)
+				log.Info("error reading message, trying again: ", res.err)
 				retry = true
 				continue
 			}
 		}
 
-		log.Event(ctx, "dhtSentMessage", ms.dht.self, ms.p, pmes)
-
-		if ms.singleMes > streamReuseTries {
-			ms.s.Close()
-			ms.s = nil
-		} else if retry {
+		if retry {
+			ms.lk.Lock()
 			ms.singleMes++
+			if ms.singleMes > streamReuseTries {
+				ms.resetSoft(rcount)
+			}
+			ms.lk.Unlock()
 		}
 
-		return mes, nil
+		return res.mes, nil
+	}
+}
+
+func (ms *messageSender) sendMessageSingle(ctx context.Context, pmes *pb.Message) error {
+	s, err := ms.dht.host.NewStream(ctx, ms.p, ProtocolDHT, ProtocolDHTOld)
+	if err != nil {
+		return err
+	}
+	defer s.Close()
+
+	w := ggio.NewDelimitedWriter(s)
+
+	err = w.WriteMsg(pmes)
+	if err != nil {
+		s.Reset()
 	}
+
+	return err
 }
 
-func (ms *messageSender) ctxReadMsg(ctx context.Context, mes *pb.Message) error {
+func (ms *messageSender) sendRequestSingle(ctx context.Context, pmes *pb.Message) (*pb.Message, error) {
+	s, err := ms.dht.host.NewStream(ctx, ms.p, ProtocolDHT, ProtocolDHTOld)
+	if err != nil {
+		return nil, err
+	}
+	defer s.Close()
+
+	r := ggio.NewDelimitedReader(s, inet.MessageSizeMax)
+	w := ggio.NewDelimitedWriter(s)
+
+	if err := w.WriteMsg(pmes); err != nil {
+		s.Reset()
+		return nil, err
+	}
+
+	mes := new(pb.Message)
+
 	errc := make(chan error, 1)
-	go func(r ggio.ReadCloser) {
+	go func() {
 		errc <- r.ReadMsg(mes)
-	}(ms.r)
+	}()
 
-	t := time.NewTimer(dhtReadMessageTimeout)
-	defer t.Stop()
+	rctx, cancel := context.WithTimeout(ctx, dhtReadMessageTimeout)
+	defer cancel()
 
 	select {
 	case err := <-errc:
-		return err
-	case <-ctx.Done():
-		return ctx.Err()
-	case <-t.C:
-		return ErrReadTimeout
+		if err != nil {
+			return nil, err
+		}
+	case <-rctx.Done():
+		return nil, rctx.Err()
+	}
+
+	return mes, nil
+}
+
+// Resets the stream unconditionally; increments the reset count.
+// Mutex must be locked.
+func (ms *messageSender) resetHard() {
+	close(ms.rch)
+	ms.s.Reset()
+	ms.s = nil
+	ms.rcount++
+}
+
+// Resets the stream only if the reset count matches the argument
+// Allows multiple read failures in batched concurrent requests with
+// only a single reset between them.
+// Mutex must be locked.
+func (ms *messageSender) resetSoft(rcount int) {
+	if rcount != ms.rcount {
```

Comment:
Shouldn't the message receiver be resetting the stream (it observed the error) instead of having the client reset? That would make the most sense, IMO. I'm *really* not a fan of using `rcount` this way.

Reply:
Hrmmmm. We do need a mechanism to unblock the receivers but reset exactly once.

Reply:
But why can't we do that from the `messageReceiver` loop? That is, when we read an error, reset the stream and return the error.

Reply:
Ok, I think I can rewrite to have the receiver close the stream and potentially eliminate the rcount contraption.

Reply:
Unfortunately If I have the client resetting, there is deadlock potential because of pipeline stalls.

Reply:
Rewrote the reset logic to eliminate the rcount contraption.